src/Analyzers/CSharp/Analyzers/NamespaceFileSync/NamespaceFileSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,177 @@
+                   s_localizableInsideMessage)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNamespaceNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace)
+                || rootNamespace is null)
+            {
+                return;
+            }
+
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || projectDir is null or { Length: 0 })
+            {
+                return;
+            }
+
+            if (context.Node is NamespaceDeclarationSyntax namespaceDecl &&
+                IsFileAndNamespaceValid(namespaceDecl, rootNamespace, projectDir, out var targetNamespace) &&
+                IsNamespaceSyncSupported(namespaceDecl, context.SemanticModel))
+            {
+                ReportDiagnostics(context, this.Descriptor, namespaceDecl, ReportDiagnostic.Warn, targetNamespace);
+            }
+        }
+
+        private static bool IsNamespaceSyncSupported(NamespaceDeclarationSyntax namespaceDeclaration, SemanticModel semanticModel)
+        {
+            var root = namespaceDeclaration.SyntaxTree.GetRoot();
+
+            // It should not be nested in other namespaces
+            var namespaceDeclCount = namespaceDeclaration
+                .AncestorsAndSelf().OfType<NamespaceDeclarationSyntax>()
+                .Count();
+            if (namespaceDeclCount > 1)
+            {
+                return false;
+            }
+
+            // It should not contain a namespace
+            var containsNamespace = namespaceDeclaration
+                 .DescendantNodes(n => n is NamespaceDeclarationSyntax)
+                 .OfType<NamespaceDeclarationSyntax>().Any();
+            if (containsNamespace)
+            {
+                return false;
+            }
+
+            // It should not contain partial classes with more than one instance in the semantic model
+            var containsPartialType = ContainsPartialTypeWithMultipleDeclarations(namespaceDeclaration, semanticModel);
+            if (containsPartialType)
+            {
+                return false;
+            }
+
+            // The current namespace should be valid
+            // TODO : can this cause a loop? This condition was used by the service before,
+            // but now could possibly add an Error severity as an analyzer
+            var isCurrentNamespaceInvalid = namespaceDeclaration.Name
+                .GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error);
```

Comment:
Maybe check for DefaultSeverity + id?

Reply:
this feels like an odd check.  i'm not sure what it's really trying to ask

Reply:
I'm not 100% sure this is needed. Even if the namespace is invalid, is it practical to suggest the correct name?

Reply:
Ping on this. Intent was to check that the namespace was valid before suggesting to fix. Thoughts on what the right behavior should be?